module/zfs/zpl_inode.c

```diff
@@ -231,6 +231,7 @@ zpl_tmpfile(struct inode *dir, struct dentry *dentry, zpl_umode_t mode)
 	cookie = spl_fstrans_mark();
 	error = -zfs_tmpfile(dir, vap, 0, mode, &ip, cr, 0, NULL);
 	if (error == 0) {
+		ip->i_mode &= ~current_umask();
```

Comment:
As I understand it, the core issue here is that unlike `.create` we can't rely of the VFS apply the umask for `.tmpfile`, so we need to handle it ourselves.  This fix makes sense but I think we should apply the mask prior to the `zpl_vap_init()` call, this should work and is more consistent with all other callers.  How about something like this (untested).

```c
        /*
         * The VFS does not apply the umask to the provided mode.
         * Therefore, it must be applied at the filesystem layer.
         */
        mode &= ~current_umask();
        zpl_vap_init(vap, dir, mode, cr);
```

Reply:
> As I understand it, the core issue here is that unlike .create we can't rely of the VFS apply the umask for .tmpfile, so we need to handle it ourselves.

yes, and even if VFS supports it at some point, I don't think this change (`mode &= ~umask`) will bother that.

> This fix makes sense but I think we should apply the mask prior to the zpl_vap_init() call, this should work and is more consistent with all other callers. How about something like this (untested).

ok, that might be better.

Reply:
Updated + rebased + tested.

Reply:
Added test case under `functional/tmpfile`.